inst/chunks/fun_normal_time.stan

```diff
@@ -1,71 +1,140 @@
+      matrix[nobs[i], nobs[i]] L_i;
+      L_i = diag_pre_multiply(sigma[begin[i]:end[i]], chol_cor[1:nobs[i], 1:nobs[i]]);
+      lp += multi_normal_cholesky_lpdf(
+        y[begin[i]:end[i]] | mu[begin[i]:end[i]], L_i
+      );
+    }
+    return lp;
+  }
+  /* multi-normal log-PDF for time-series covariance structures
+   * in Cholesky parameterization and assuming homogoneous variances
+   * allows for flexible correlation matrix subsets
+   * Deviating Args:
+   *   Jtime: array of time indices per group
+   * Returns:
+   *   sum of the log-PDF values of all observations
+   */
+  real normal_time_hom_flex_lpdf(vector y, vector mu, real sigma, matrix chol_cor,
+                                 int[] nobs, int[] begin, int[] end, int[,] Jtime) {
+    real lp = 0.0;
+    int I = size(nobs);
+    int has_lp[I] = rep_array(0, I);
+    int i = 1;
+    matrix[rows(chol_cor), cols(chol_cor)] Cor;
+    Cor = multiply_lower_tri_self_transpose(chol_cor);
+    while (sum(has_lp) != I) {
+      int iobs[nobs[i]] = Jtime[i, 1:nobs[i]];
+      int lp_terms[I-i+1] = rep_array(0, I-i+1);
+      matrix[nobs[i], nobs[i]] L_i;
+      if (is_equal(iobs, sequence(1, rows(chol_cor)))) {
+        // all timepoints are present in this group
+        L_i = chol_cor;
+      } else {
+        // arbitrary subsets cannot be taken on chol_cor directly
+        L_i = cholesky_decompose(Cor[iobs, iobs]);
       }
-      lp[i] = multi_normal_cholesky_lpdf(
-        y[begin[i]:end[i]] | mu[begin[i]:end[i]], L
+      L_i = sigma * L_i;
+      has_lp[i] = 1;
+      lp_terms[1] = 1;
+      // find all additional groups where we have the same timepoints
+      for (j in (i+1):I) {
+        if (has_lp[j] == 0 && is_equal(Jtime[j], Jtime[i]) == 1) {
+          has_lp[j] = 1;
+          lp_terms[j-i+1] = 1;
+        }
+      }
+      // vectorize the log likelihood by stacking the vectors
+      lp += multi_normal_cholesky_lpdf(
+        stack_vectors(y, nobs[i], lp_terms, begin[i:I], end[i:I]) |
+        stack_vectors(mu, nobs[i], lp_terms, begin[i:I], end[i:I]), L_i
       );
+      while (has_lp[i] == 1 && i != I) {
+        i += 1;
+      }
+    }
+    return lp;
+  }
+  /* multi-normal log-PDF for time-series covariance structures
+   * in Cholesky parameterization and assuming heterogenous variances
+   * allows for flexible correlation matrix subsets
+   * Deviating Args:
+   *   sigma: residual standard deviation vectors
+   *   Jtime: array of time indices per group
+   * Returns:
+   *   sum of the log-PDF values of all observations
+   */
+  real normal_time_het_flex_lpdf(vector y, vector mu, vector sigma, matrix chol_cor,
+                                 int[] nobs, int[] begin, int[] end, int[,] Jtime) {
+    real lp = 0.0;
+    int I = size(nobs);
+    int has_lp[I] = rep_array(0, I);
+    int i = 1;
+    matrix[rows(chol_cor), cols(chol_cor)] Cor;
+    Cor = multiply_lower_tri_self_transpose(chol_cor);
+    while (sum(has_lp) != I) {
```

Comment:
we could avoid the `sum(has_lp)` check here, which is a sum over all groups. Instead we could count i to I+1 and abort if that value is reached... not a huge time saver though as we likely won't apply this to huge data, but we can avoid the repeated sum over `has_lp` (which we still need to flag which terms have been collected into `lp` as we go).

Reply:
Does Stan evaluate `y` of `x && y` is `x`already is FALSE?

Reply:
Asked differently, can I safely use 

```Stan
while (i <= I && has_lp[i] == 1) {
        i += 1;
      }
```

as this will mean i becoming I+1 at some point

Reply:
1) yes, that's correct, see https://mc-stan.org/docs/2_25/functions-reference/logical-functions.html#boolean-operator-short-circuiting

2) yes, this logic should work.

Reply:
Tried it out and seems to work.